ArduPlane/soaring.cpp

```diff
@@ -0,0 +1,74 @@
+#include "Plane.h"
+
+/*
+*  ArduSoar support function
+*
+*  Peter Braswell, Samuel Tabor, Andrey Kolobov, and Iain Guilliard
+*/
+void Plane::update_soaring() {
+    
+    if (!soaring_controller.is_active()) {
+        return;
+    }
+    
+    soaring_controller.update_vario();
+
+    // Check for throttle suppression change.
+    switch (control_mode){
+    case AUTO:
+        soaring_controller.suppress_throttle();
+        break;
+    case FLY_BY_WIRE_B:
+        if (!soaring_controller.suppress_throttle()) {
+            set_mode(RTL, MODE_REASON_SOARING_FBW_B_WITH_MOTOR_RUNNING);
+        }
+        break;
+    case LOITER:
+        // Do nothing. We will switch back to auto/rtl before enabling throttle.
+        break;
+    default:
+        // This does not affect the throttle since suppressed is only checked in the above three modes. 
+        // It ensures that the soaring always starts with throttle suppressed though.
+        soaring_controller.set_throttle_suppressed(true);
+        break;
+    }
+    // Nothing to do if we are in powered flight
+    if (!soaring_controller.get_throttle_suppressed() && aparm.throttle_max > 0) {
+        return;
+    }
+
+    switch (control_mode){
+    case AUTO:
+    case FLY_BY_WIRE_B:
+        // Test for switch into thermalling mode
+        if (soaring_controller.check_thermal_criteria()) {
+            GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "Thermal detected, entering loiter\n");
+            set_mode(LOITER, MODE_REASON_SOARING_THERMAL_DETECTED);
+        }
+        break;
+    case LOITER:
+        if (soaring_controller.check_init_thermal_criteria()) {

```

Comment:
There's a change to the logic here that I think disables update of the navigation waypoint until after min_thermal_s has passed. Is this intentional?
I have found that the initial stages of thermal centreing, when the state estimate variances are still high, is the most important for successful estimation, and allowing the aircraft to fly to the estimated centre location improves performance in this regard.